staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -165,6 +229,126 @@ func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+		return false, v1.LabelFailureDomainBetaZone, v1.LabelFailureDomainBetaRegion
+	}
+
+	// If nothing is in the NodeAfinity, we should check pv labels
+	_, zoneGA = pv.Labels[v1.LabelTopologyZone]
+	_, regionGA = pv.Labels[v1.LabelTopologyRegion]
+	if zoneGA || regionGA {
+		// NodeAffinity not exist, GA label exists
+		return true, v1.LabelTopologyZone, v1.LabelTopologyRegion
+	}
+
+	// If GA label not exists, check beta version
+	_, zoneBeta = pv.Labels[v1.LabelFailureDomainBetaZone]
+	_, regionBeta = pv.Labels[v1.LabelFailureDomainBetaRegion]
+	if zoneBeta || regionBeta {
+		// Beta label exists, NodeAffinity not exist, GA label not exists
+		return false, v1.LabelFailureDomainBetaZone, v1.LabelFailureDomainBetaRegion
+	}
+
+	// No labels or NodeAffinity exist, default to GA version
+	return true, v1.LabelTopologyZone, v1.LabelTopologyRegion
+}
+
+// TopologyKeyExist checks if a certain key exists in a VolumeNodeAffinity
+func TopologyKeyExist(key string, vna *v1.VolumeNodeAffinity) bool {
+	if vna == nil || vna.Required == nil || vna.Required.NodeSelectorTerms == nil || len(vna.Required.NodeSelectorTerms) == 0 {
+		return false
+	}
+
+	for _, nodeSelectorTerms := range vna.Required.NodeSelectorTerms {
+		nsrequirements := nodeSelectorTerms.MatchExpressions
+		for _, nodeSelectorRequirement := range nsrequirements {
+			if nodeSelectorRequirement.Key == key {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+type regionTopologyHandler func(pv *v1.PersistentVolume) error
+
+// translateTopologyFromCSIToInTree translate a CSI topology to
+// Kubernetes topology and add topology labels to it. Note that this function
+// will only work for plugin with single topologyKey. If a plugin has
+// more than one topologyKey, it will need to be processed separately
+// by the plugin.
+// regionTopologyHandler is a function to generate region topology NodeAffinity
+// If regionTopologyHandler is nil, no region NodeAffinity will be added
+// 1. Remove any existing CSI topologyKey from NodeAffinity
+// 2. Add Kubernetes Topology in the NodeAffinity if it does not exist
+//    2.1 Try to use CSI topology values to recover the Kubernetes topology first
+//    2.2 If CSI topology values does not exist, try to use PV label
+// 3. Add Kubernetes Topology labels(zone and region)
+func translateTopologyFromCSIToInTree(pv *v1.PersistentVolume, topologyKey string, regionTopologyHandler regionTopologyHandler) error {
+
+	_, zoneLabel, regionLabel := getTopologyLabel(pv)
+
+	// 1. Replace all CSI topology to Kubernetes Zone label
+	err := replaceTopology(pv, topologyKey, zoneLabel)
+	if err != nil {
+		return fmt.Errorf("Failed to replace CSI topology to Kubernetes topology, error: %v", err)
+	}
+
+	// 2. Take care of region topology if a regionTopologyHandler is passed
+	if regionTopologyHandler != nil {
+		err = regionTopologyHandler(pv)
+		if err != nil {
+			// let's make less strict on this one. Even if there is an error in the region processing, just ignore it
```

Comment:
One thing I want to call out in this commit is that I silently ignore any error from the  regionTopologyHandler just want to make the function less error-prone